Provide Promscale upgrading guide (#1136)
timescale/docs: promscale/installation/upgrade.md

# Upgrade Promscale

> Promscale consists of the Promscale Connector, and the Promscale PostgreSQL
> extension. When you upgrade your Promscale installation, you need to check both

Reviewer: ```suggestion
Promscale consists of the Promscale Connector, and the Promscale Database. The Promscale Database
is PostgreSQL with the TimescaleDB and the Promscale extensions. When you upgrade your Promscale installation, you need to check both
```

> components. This section contains important information about upgrading.

Reviewer: ```suggestion
the Connector and the Database. This section contains important information about upgrading.
```

> The process for updating your Promscale installation is:

Reviewer: I think we need a process for when only the extension needs to be upgraded and maybe a separate process for when PostgreSQL and/or TimescaleDB need to be upgraded. Ideally we can point to Postgres and/or TimescaleDB documentation for those upgrades.

> 1.  Check that the most recent required version of the Promscale extension is
>     available.

Reviewer: ```suggestion
1.  Check that the most recent required versions of PostgreSQL, TimescaleDB and the
Promscale extension are available.
```

Reviewer: Can we also provide instructions on how to check the versions of PostgreSQL, TimescaleDB and Promscale?

For TimescaleDB and Promscale this would be. Show current version installed:

`SELECT extname, extversion FROM pg_extension WHERE extname IN ('timescaledb', 'promscale');
`
Show available versions:

`SELECT name, version FROM pg_available_extension_versions WHERE name IN ('timescaledb', 'promscale');`

Reviewer: To get the PostgreSQL version:
`SHOW server_version;`

> 1.  Stop all Promscale Connector instances connected to the database.
> 1.  Perform the upgrade.
> 1.  Start the new version of the Promscale Connector, and wait for it to begin

Reviewer: ```suggestion
1.  Stop all Promscale Connector instances connected to the database.
1.  Start the new version of the Promscale Connector, and wait for it to begin
```

Reviewer: There is no upgrade to be performed since the upgrade happens automatically once the Promscale Connector is installed.

> 1.  Check that the most recent required version of the Promscale extension is
>     available.
> 2.  Stop all Promscale Connector instances connected to the database.
> 3.  Perform the upgrade.
> 4.  Start the new version of the Promscale Connector, and wait for it to begin
>     ingesting data.
> 5.  Start all other Promscale Connector instances.

Reviewer 2: ```suggestion
1.  Check that the most recent required version of the Promscale extension is
    available.
1.  Stop all Promscale Connector instances connected to the database.
1.  Perform the upgrade.
1.  Start the new version of the Promscale Connector, and wait for it to begin
    ingesting data.
1.  Start all other Promscale Connector instances.
```

> ## Upgrade to Promscale&nbsp;0.11.0

Author: How/when do you decide when to use `&nbsp;`?

Author: It seems as though using `&nbsp;` in a header kills whatever follows the space.

Reviewer 2: It should be a non-breaking space between the product name and version number, but if it's breaking in the header we'll need to make it a regular space, I guess.

> ## Upgrade to Promscale&nbsp;0.11.0

Reviewer: ```suggestion
## Upgrade to Promscale 0.11.0
```

Promscale&nbsp;0.11.0 contains significant changes, and the upgrade drops any
previous tracing data you have stored. Make a backup of your installation, and
test the new version, before you go ahead with this upgrade.

Upgrading to Promscale&nbsp;0.11.0 could take up to 5-10 minutes on very large
databases. During the upgrade, Promscale is not able to ingest or query data.

The upgrade process modifies a lot of database objects. On large deployments,
you might need to increase the PostgreSQL `max_locks_per_transaction` parameter
before you start the upgrade.

<highlight type="warning">

> When you upgrade to Promscale&nbsp;0.11.0, all previous tracing data is

Author: Do we purposely duplicate the content here with that from three paragraphs above?

Reviewer 2: Yes. People have a nasty tendency to not read admonitions, so in the case of really important ones, I like to duplicate the content.

dropped. Make a backup of your installation, and test the new version
before you upgrade.
</highlight>

> You must have installed TimescaleDB&nbsp;2.6.1 or later.

Reviewer: ```suggestion
You must have installed PostgreSQL 12 or later and TimescaleDB 2.6.1 or later.
```

The Promscale extension is now mandatory. You must be using version 0.5.0 or
later. Before you start the upgrade, make sure that you have made the most
recent version of the Promscale extension available to PostgresSQL. You must use
the Promscale Connector to update the extension, do not use the
`ALTER EXTENSION promscale UPDATE;` command. For more information on how to do

> this, see the [Promscale installation instructions](install-promscale) for your

Author: I broke this 😢 
```suggestion
this, see the [Promscale installation instructions][install-promscale] for your
```

preferred installation method.

Upgrading to Promscale&nbsp;0.11.0 requires creating a lock on your tables. If
you have been using Promscale for a while, and you have a large number of
metrics, this could result in locking a large number of tables. If PostgreSQL
has not been correctly configured for this, the process could fail with an error
like this:

```yml
ERROR: out of shared memory (SQLSTATE 53200)`.
```

This error occurs because PostgreSQL has a limited amount of shared memory
available to store locks on objects. You can increase the amount of shared
memory available by increasing the `max_locks_per_transaction` parameter. When
you have completed the upgrade, you can set the value back to what it was before
the upgrade. For more information about tuning this parameter, see

> [troubleshooting Promscale][max_locks_config].

Author: ```suggestion
[troubleshooting Promscale][max-locks-config].
```

<procedure>

### Upgrading to Promscale&nbsp;0.11.0

> <!--- Insert procedure here -->

Author: I'm not sure what should actually go in here? Should we repeat the general install instructions above?

Author: Ah, I see. Should we not have one generic "how to upgrade" process at the top of this page? Do we need to repeat the instructions again for version 0.11.0?

Reviewer 2: I guess that depends on if the process is going to be different with each version or not?

Author: It's hard to predict the future, but I don't think so. If it is different in a future version then we could update these docs to account for that?

tracing data is dropped. Make a backup of your installation, and test the new
version before you upgrade. </highlight>

<procedure>

### Upgrading to Promscale 0.11.0
To upgrade to Promscale 0.11.0 you must have installed PostgreSQL 12 or later,
TimescaleDB 2.6.1 or later, and Promscale extension version 0.5.0 or later. 

1. Check the version of PostgreSQL that is installed:
   ```sql
   SHOW server_version;
   ``` 
1. Check the version of TimescaleDB and Promscale that is installed:
   ```sql
   SELECT extname, extversion FROM pg_extension WHERE extname IN ('timescaledb', 'promscale');
   ```
   For information about upgrading TimescaleDB and PostgreSQL, see [Update TimescaleDB][update-timescaledb] and [Upgrade PostgreSQL][upgrade-postgresql].
1. Stop all Promscale Connector instances connected to the database.
1. Update the Promscale extension on one instance using the Promscale Connector.
   For more information, see the [Promscale installation
   instructions][install-promscale] for various installation method.
   <highlight type="note"> Do not use the `ALTER EXTENSION promscale UPDATE;` command. to update the extension.
   </highlight>

> 1. Start one instance with the the latest version of the Promscale Connector. The migration happens automatically. After the migration is completed, upgrade the remaining Promscale Connector instances.

Reviewer: ```suggestion
1. Start one instance with the latest version of the Promscale Connector. The migration happens automatically. After the migration is completed, upgrade the remaining Promscale Connector instances.
```

# Upgrade Promscale
Promscale consists of the Promscale Connector, and the Promscale Database. The
Promscale Database is PostgreSQL with the TimescaleDB and the Promscale
extensions. When you upgrade your Promscale installation, you need to check both
the Connector and the Database. This section contains important information
about upgrading.

The process for updating your Promscale installation is:

1.  Check that the most recent required versions of PostgreSQL, TimescaleDB and
the Promscale extension are available.
1.  Stop all Promscale Connector instances connected to the database.

> 1.  Start the new version of the Promscale Connector, and wait for it to begin
>     ingesting data.

Reviewer: ```suggestion
1.  Start one instance with the new version of the Promscale Connector.
```

tracing data is dropped. Make a backup of your installation, and test the new
version before you upgrade. </highlight>

<procedure>

### Upgrading to Promscale 0.11.0
To upgrade to Promscale 0.11.0 you must have installed PostgreSQL 12 or later,
TimescaleDB 2.6.1 or later, and Promscale extension version 0.5.0 or later. 

1. Check the version of PostgreSQL that is installed:
   ```sql
   SHOW server_version;
   ``` 
1. Check the version of TimescaleDB and Promscale that is installed:
   ```sql
   SELECT extname, extversion FROM pg_extension WHERE extname IN ('timescaledb', 'promscale');
   ```
   For information about upgrading TimescaleDB and PostgreSQL, see [Update TimescaleDB][update-timescaledb] and [Upgrade PostgreSQL][upgrade-postgresql].
1. Stop all Promscale Connector instances connected to the database.
1. Update the Promscale extension on one instance using the Promscale Connector.
   For more information, see the [Promscale installation
   instructions][install-promscale] for various installation method.
   <highlight type="note"> Do not use the `ALTER EXTENSION promscale UPDATE;` command. to update the extension.
   </highlight>

> 1. Start the latest version of the Promscale Connector, and wait for it to begin
>    ingesting data. The migration happens automatically. After the migration is completed, upgrade the remaining Promscale instances.

Reviewer: ```suggestion
1. Start one instance with the the latest version of the Promscale Connector. The migration happens automatically. After the migration is completed, upgrade the remaining Promscale Connector instances.
```
